rules/do-you-use-loggermessage-in-net.md/rule.md

```diff
@@ -0,0 +1,118 @@
+---
+seoDescription: Learn how to optimize logging in .NET applications using the LoggerMessage class. Discover the performance benefits, such as reduced object allocations and efficient message parsing, and get practical tips on how to implement this pattern for better logging efficiency and structured logs in high-performance scenarios.
+type: rule
+title: Do you use LoggerMessage in .NET?
+uri: do-you-use-loggermessage-in-net
+authors:
+  - title: Tom Iwainski
+    url: https://www.ssw.com.au/people/thomas-iwainski
+related: []
+created: 2024-11-05T12:52:08.000Z
+archivedreason: null
+guid: 2feb782f-0cee-4022-95d9-f44b9f739b90
+---
+
+Logging is a critical component in modern applications, but it can easily introduce performance overhead.
+
+The LoggerMessage class in .NET offers a solution to reduce the performance costs associated with logging. By utilizing this pattern, developers can optimize logging with fewer object allocations and reduced overhead.
+
+<!--endintro-->
+
+## Key performance benefits of LoggerMessage
+
+1. **Reduced Object Allocations**: Traditional logger extension methods often require "boxing" of value types (e.g., integers or dates) into objects. LoggerMessage avoids this by using static Action fields and strongly typed parameters, eliminating the need for boxing and reducing memory overhead.
+
+2. **Efficient Message Parsing**: Logger extension methods parse the log message template each time a log entry is made. In contrast, LoggerMessage only parses the template once when the message is defined, leading to fewer computations and faster execution during logging.
+
+3. **Structured Logging**: With LoggerMessage, log templates can include placeholders that are filled with specific log data, such as Item or DateTime. This makes it easier to integrate structured logging with logging frameworks like Serilog, which can enrich logs with meaningful event names.
+
+## How to use LoggerMessage
+
+To use the LoggerMessage pattern, you define a static Action delegate for logging a message. The LoggerMessage.Define method is used to create these delegates with parameters such as:
+
+- Log Level (e.g., LogLevel.Information)
+- Event ID (for tracking specific events)
+- Message Template (a named format string with placeholders)
+
+For example:
+
+Each log message is represented as an Action delegate.
+
+```csharp
+private static readonly Action<ILogger, WorkItem, Exception> s_processingPriorityItem =
```

Comment:
These examples aren't right. Have a look at this documentation:

https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator

You need to use the `[LoggerMessage]` attribute, for the source generators to work.

You should also mention that source generators are what makes this work so well.

Reply:
Ok, seems we were talking about 2 different things?

https://learn.microsoft.com/en-us/dotnet/core/extensions/high-performance-logging

Reply:
The source generated loggers generate logger messages under the hood (I think, but you may need to double check).

I think we should be directly developers to use the source generated logs that should gain all the perf benefits of the Logger message stuff you mention in the rule.